### PR TITLE
[Numpy] FFI for linalg ops

### PR DESCRIPTION
**Pull request by sjtuWangDing**
## Description ##
FFI for linalg inv/pinv/eigvals/eigvalsh/solve/tensorsolve/tensorinv/cholesky

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/api/operator/numpy/linalg/np_eigvals.cc [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  const nnvm::Op* op = Op::Get("_npi_eigvalsh");
+  nnvm::NodeAttrs attrs;
+  op::EigvalshParam param;
+  param.UPLO = *((args[1].operator std::string()).c_str());
+  attrs.parsed = param;
+  attrs.op = op;
+  SetAttrDict<op::EigvalshParam>(&attrs);
+  int num_outputs = 0;
+  NDArray* inputs[] = {args[0].operator mxnet::NDArray*()};
+  auto ndoutputs = Invoke(op, &attrs, 1, inputs, &num_outputs, nullptr);
```
Comment (hzfan) [inline src/api/operator/numpy/linalg/np_eigvals.cc]: `1` is a magic number, please use `int num_inputs=1`

File: src/api/operator/numpy/linalg/np_pinv.cc [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  op::PinvScalarRcondParam param;
+  nnvm::NodeAttrs attrs;
+  param.rcond = args[1].operator double();
+  param.hermitian = args[2].operator bool();
+  attrs.parsed = param;
+  attrs.op = op;
+  SetAttrDict<op::PinvScalarRcondParam>(&attrs);
+  int num_outputs = 0;
+  NDArray* inputs[] = {args[0].operator mxnet::NDArray*()};
+  auto ndoutputs = Invoke(op, &attrs, 1, inputs, &num_outputs, nullptr);
```
Comment (hzfan) [inline src/api/operator/numpy/linalg/np_pinv.cc]: Same

File: src/api/operator/numpy/linalg/np_pinv.cc [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  const nnvm::Op* op = Op::Get("_npi_pinv");
+  op::PinvParam param;
+  nnvm::NodeAttrs attrs;
+  param.hermitian = args[2].operator bool();
+  attrs.parsed = param;
+  attrs.op = op;
+  SetAttrDict<op::PinvParam>(&attrs);
+  int num_outputs = 0;
+  NDArray* inputs[] = {args[0].operator mxnet::NDArray*(), args[1].operator mxnet::NDArray*()};
+  auto ndoutputs = Invoke(op, &attrs, 2, inputs, &num_outputs, nullptr);
```
Comment (hzfan) [inline src/api/operator/numpy/linalg/np_pinv.cc]: Same

File: src/api/operator/numpy/linalg/np_potrf.cc [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  const nnvm::Op* op = Op::Get("_npi_cholesky");
+  nnvm::NodeAttrs attrs;
+  op::LaCholeskyParam param;
+  param.lower = args[1].operator bool();
+  attrs.parsed = param;
+  attrs.op = op;
+  SetAttrDict<op::LaCholeskyParam>(&attrs);
+  int num_outputs = 0;
+  NDArray* inputs[] = {args[0].operator mxnet::NDArray*()};
+  auto ndoutputs = Invoke(op, &attrs, 1, inputs, &num_outputs, nullptr);
```
Comment (hzfan) [inline src/api/operator/numpy/linalg/np_potrf.cc]: Same.

File: src/api/operator/numpy/linalg/np_tensorinv.cc [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  const nnvm::Op* op = Op::Get("_npi_tensorinv");
+  nnvm::NodeAttrs attrs;
+  op::TensorinvParam param;
+  param.ind = args[1].operator int();
+  attrs.parsed = param;
+  attrs.op = op;
+  SetAttrDict<op::TensorinvParam>(&attrs);
+  int num_outputs = 0;
+  NDArray* inputs[] = {args[0].operator mxnet::NDArray*()};
+  auto ndoutputs = Invoke(op, &attrs, 1, inputs, &num_outputs, nullptr);
```
Comment (hzfan) [inline src/api/operator/numpy/linalg/np_tensorinv.cc]: Same

File: src/api/operator/numpy/linalg/np_tensorsolve.cc [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    } else {
+      param.a_axes = Tuple<int>(args[2].operator ObjectRef());
+    }
+  }
+  attrs.parsed = param;
+  attrs.op = op;
+  SetAttrDict<op::TensorsolveParam>(&attrs);
+  int num_outputs = 0;
+  NDArray* inputs[] = {args[0].operator mxnet::NDArray*(), args[1].operator mxnet::NDArray*()};
+  auto ndoutputs = Invoke(op, &attrs, 2, inputs, &num_outputs, nullptr);
```
Comment (hzfan) [inline src/api/operator/numpy/linalg/np_tensorsolve.cc]: Same

File: python/mxnet/numpy/linalg.py [hunk 7]
```diff
@@ -232,7 +232,7 @@ def svd(a):
     return _mx_nd_np.linalg.svd(a)
 
 
-def cholesky(a):
+def cholesky(a, lower=True):
```
Comment (hzfan) [inline python/mxnet/numpy/linalg.py]: Seems official numpy does not support lower?
https://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.cholesky.html

File: src/api/operator/numpy/linalg/np_pinv.cc [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+/*!
+ * \file np_pinv.cc
+ * \brief Implementation of the API of functions in src/operator/numpy/linalg/np_pinv.cc
+ */
+
+#include <mxnet/api_registry.h>
+#include <mxnet/runtime/packed_func.h>
+#include "../../utils.h"
+#include "../../../../operator/numpy/linalg/np_pinv-inl.h"
```
Comment (hzfan) [inline src/api/operator/numpy/linalg/np_pinv.cc]: `linalg` is a bit different from ordinary ops. Could you have a seperate file like `np_pinv.h` for each linalg ops? This header file only needs to contain `#include "../../../../operator/numpy/linalg/np_pinv-inl.h"`. After that, we remove this line here and place `#include "./np_pinv.h` at the top of all these `include` statements.

File: python/mxnet/numpy/linalg.py (hunk 7 above)
Comment (sjtuWangDing) [reply python/mxnet/numpy/linalg.py]: Official numpy not support 'lower'. At the backend, the implementation of 'cholesky' is to call 'potrf' in the file la_op-inl.h, and it needs a parameter 'lower'. I will use this parameter directly in the return statement, just like '_api_internal.cholesky(a, True)' and interface will be the same as official numpy.

**Comment by hzfan** [comment]
@mxnet-bot run ci [centos-cpu, windows-gpu]

**Comment by sjtuWangDing** [comment]
run ci [centos-cpu, windows-gpu]

**Comment by sjtuWangDing** [comment]
@mxnet-bot run ci [centos-cpu, windows-gpu]

**Comment by sjtuWangDing** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by leezu** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by sjtuWangDing** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by yzhliu** [comment]
Thanks @sjtuWangDing @hzfan

**Pull request merged**